ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java

```diff
@@ -166,96 +161,34 @@ void testConnectivity() {
-   * @param query the DocumentClient query to execute, wrapped in a {@link Callable}.
-   * @param timer the {@link Timer} to use to record execution time (excluding waiting).
-   * @return the {@link FeedResponse} returned by the query, if successful.
-   * @throws DocumentClientException if Cosmos returns a different error status, or if the retry limit is reached.
-   */
-  private FeedResponse<Document> retryQueryWithThrottling(Callable<FeedResponse<Document>> query, Timer timer)
-      throws DocumentClientException {
-    return (FeedResponse<Document>) retryWithThrottling(query, timer);
-  }
-
-  /**
-   * Run the supplied DocumentClient action. If CosmosDB returns status 429 (TOO_MANY_REQUESTS),
-   * retry after the requested wait period, up to the configured retry limit.
-   * @param action the DocumentClient action to execute, wrapped in a {@link Callable}.
-   * @param timer the {@link Timer} to use to record execution time (excluding waiting).
-   * @return the {@link Object} returned by the action, if successful.
-   * @throws DocumentClientException if Cosmos returns a different error status, or if the retry limit is reached.
-   */
-  private Object retryWithThrottling(Callable<? extends Object> action, Timer timer) throws DocumentClientException {
-    int count = 0;
-    long waitTime = 0;
-    do {
-      try {
-        waitForMs(waitTime);
-        Timer.Context docTimer = timer.time();
-        Object response = executeCosmosAction(action);
-        docTimer.stop();
-        return response;
-      } catch (DocumentClientException dex) {
-        // Azure tells us how long to wait before retrying.
-        if (dex.getStatusCode() == HTTP_TOO_MANY_REQUESTS) {
-          waitTime = dex.getRetryAfterInMilliseconds();
-          azureMetrics.retryCount.inc();
-          logger.debug("Got {} from Cosmos, will wait {} ms before retrying.", HTTP_TOO_MANY_REQUESTS, waitTime);
-        } else {
-          // Something else, not retryable.
-          throw dex;
-        }
-      } catch (Exception e) {
-        azureMetrics.documentErrorCount.inc();
-        throw new RuntimeException("Exception calling action " + action, e);
-      }
-      count++;
-    } while (count <= maxRetries);
-    azureMetrics.retryCount.dec(); // number of retries, not total tries
-    azureMetrics.documentErrorCount.inc();
-    String message = "Max number of retries reached while retrying with action " + action;
-    throw new RuntimeException(message);
-  }
-
   /**
    * Utility method to call a Cosmos method and extract any nested DocumentClientException.
    * @param action the action to call.
    * @return the result of the action.
-   * @throws Exception
+   * @throws DocumentClientException
    */
-  private Object executeCosmosAction(Callable<? extends Object> action) throws Exception {
+  private ResourceResponse<Document> executeCosmosAction(Callable<? extends Object> action, Timer timer)
```

Comment:
would making the parameter a `Callable<? extends ResourceResponse<? extends Document>>`, or `Callable<ResourceResponse<Document>>` work?

Reply:
done.